File: leetcode/96.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
   public:
    vector<int> table;
    void creatTable(int n) {
        n++;
        table.assign(n, 0);
        table[0] = 1;
        table[1] = 1;
        for (int i = 2; i < n; i++) {
            for (int j = 0; j < i; j++) {
                table[i] += table[j] * table[i - 1 - j];
            }
        }
    }
    int numTrees(int n) {
        creatTable(n);
        return table[n];
    }
};

File: leetcode/96.js
/**
 * @param {number} n
 * @return {number}
 */
var numTrees = function (n) {
    function recur() {
        let table = [1, 1, 2, 5];
        let resFunc = function (n) {
            if (n < table.length) {
                return table[n];
            } else {
                let res = 0;
                for (let i = 1; i <= n; i++) {
                    res += resFunc(i - 1) * resFunc(n - i);
                }
                table[n] = res;
                return res;
            }
        }
        return resFunc;
    }
    return recur()(n);
}


// console.log(numTrees(30)) ;